internal/model/keys.go

```diff
@@ -2,85 +2,157 @@
 package model
 
 import (
+	"crypto/rand"
 	"database/sql"
 	"database/sql/driver"
 	"fmt"
+	"math/big"
 	"reflect"
-	"regexp"
+	"strconv"
 	"strings"
-
-	"github.com/google/uuid"
-	"github.com/oklog/ulid/v2"
 )
 
-func NewRunnerKey() RunnerKey                      { return RunnerKey(ulid.Make()) }
-func ParseRunnerKey(key string) (RunnerKey, error) { return parseKey[RunnerKey](key) }
+func NewRunnerKey(hostname string, port string) RunnerKey {
+	suffix, err := rand.Int(rand.Reader, big.NewInt(10000))
+	if err != nil {
+		panic(err)
+	}
+	return keyType[runnerKey]{
+		Hostname: hostname,
+		Port:     port,
+		Suffix:   int(suffix.Int64()),
+	}
+}
+func NewLocalRunnerKey(suffix int) RunnerKey {
+	return keyType[runnerKey]{
+		Suffix: suffix,
+	}
+}
+func ParseRunnerKey(key string) (RunnerKey, error) { return parseKey[RunnerKey](key, true) }
 
 type runnerKey struct{}
 type RunnerKey = keyType[runnerKey]
 
-func NewControllerKey() ControllerKey                      { return ControllerKey(ulid.Make()) }
-func ParseControllerKey(key string) (ControllerKey, error) { return parseKey[ControllerKey](key) }
+func NewControllerKey(hostname string, port string) ControllerKey {
+	suffix, err := rand.Int(rand.Reader, big.NewInt(10000))
```

Comment:
Let's use 4 bytes (4294967296) of randomness so it's more consistent with DeploymentName.